blocks/fileio/include/gnuradio-4.0/fileio/BasicFileIo.hpp

```diff
@@ -0,0 +1,307 @@
+        }
+    }
+
+    return deletedFiles;
+}
+
+} // namespace detail
+
+enum class FileMode { override, append, multi };
+
+template<typename T>
+struct BasicFileSink : public gr::Block<BasicFileSink<T>> {
+    using Description = Doc<R""(A sink block for writing a stream to a binary file.
+The file can be played back using a 'BasicFileSource' or read by any program that supports binary files (e.g. Python, C, C++, MATLAB).
+For complex types, the binary file contains [float, double]s in IQIQIQ order. No metadata is included with the binary data.)
+Important: this implementation assumes a host-order, CPU architecture specific byte order!)"">;
+    template<typename U, gr::meta::fixed_string description = "", typename... Arguments>
+    using A = gr::Annotated<U, description, Arguments...>; // optional shortening
+
+    PortIn<T> in;
+
+    A<std::string, "file name", Doc<"base filename, prefixed if ">, Visible>              file_name;
+    FileMode                                                                              _mode              = FileMode::override;
+    A<std::string, "mode", Doc<"mode: \"override\", \"append\", \"multi\"">, Visible>     mode               = std::string(magic_enum::enum_name(_mode));
+    A<gr::Size_t, "max bytes per file", Doc<"max bytes per file, 0: infinite ">, Visible> max_bytes_per_file = 0U;
+
+    std::size_t   _totalBytesWritten{0UZ};
+    std::size_t   _totalBytesWrittenFile{0UZ};
+    std::ofstream _file;
+    std::size_t   _fileCounter{0UZ};
+    std::string   _actualFileName;
+
+    void settingsChanged(const property_map& /*oldSettings*/, const property_map& /*newSettings*/) {
+        _mode = magic_enum::enum_cast<FileMode>(mode, magic_enum::case_insensitive).value_or(_mode);
+        if (lifecycle::isActive(this->state())) {
+            closeFile();
+            openNextFile();
+        }
+    }
+
+    void start() {
+        _totalBytesWritten = 0UZ;
+        openNextFile();
+    }
+
+    void stop() { closeFile(); }
+
+    [[nodiscard]] constexpr work::Status processBulk(ConsumableSpan auto& dataIn) {
+        std::size_t nBytesMax = dataIn.size() * sizeof(T);
+        if (max_bytes_per_file.value != 0U) {
+            nBytesMax = std::min(nBytesMax, static_cast<std::size_t>(max_bytes_per_file.value - _totalBytesWrittenFile));
+        }
+        _file.write(reinterpret_cast<const char*>(dataIn.data()), static_cast<std::streamsize>(nBytesMax));
+        if (!dataIn.consume(nBytesMax / sizeof(T))) {
+            throw gr::exception("could not consume input samples");
+        }
+
+        if (!_file) {
+            throw gr::exception(fmt::format("failed to write to file '{}'.", _actualFileName));
+        }
+
+        _totalBytesWritten += nBytesMax;
+        _totalBytesWrittenFile += nBytesMax;
+        if (max_bytes_per_file.value != 0U && _totalBytesWrittenFile >= max_bytes_per_file.value) {
+            closeFile();
+            openNextFile();
+        }
```

Comment:
If this `processBulk()` call is the last one (there is no more data after it) and a file was just filled up, this will create a new empty file which will never have data. It seems better to do this check at the beginning of `processBulk()` rather than at the end: if we cannot write any data because the current file is full, we create a new file and use that instead. That logic guarantees that we don't end up with empty files.

Reply:
changed it.